lib/integrations/azureoidc/enterprise_app.go

```diff
@@ -0,0 +1,168 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package azureoidc
+
+import (
+	"context"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	msgraphsdk "github.com/microsoftgraph/msgraph-sdk-go"
+	"github.com/microsoftgraph/msgraph-sdk-go/applicationtemplates"
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+	"github.com/microsoftgraph/msgraph-sdk-go/serviceprincipals"
+)
+
+// A special application template ID in Microsoft Graph, equivalent to the "create your own application" option in Azure portal.
+// Only non-gallery apps ("Create your own application" option in the UI) are allowed to use SAML SSO,
+// hence we use this template.
+// Ref: https://learn.microsoft.com/en-us/graph/api/applicationtemplate-instantiate
+const nonGalleryAppTemplateID = "8adf8e6e-67b2-4cf2-a259-e3dc5476c621"
+
+// A list of Microsoft Graph permissions ("app roles") for directory sync as performed by the Entra ID plugin.
+// Ref: https://learn.microsoft.com/en-us/graph/permissions-reference
+var appRoles = []string{
+	// Application.Read.All
+	"9a5d68dd-52b0-4cc2-bd40-abcf44ac3a30",
+	// Directory.Read.All
+	"7ab1d382-f21e-4acd-a863-ba3e13f7da61",
+	// Policy.Read.All
+	"246dd0d5-5bd0-4def-940b-0421030a5b68",
+}
+
+// SetupEnterpriseApp sets up an Enterprise Application in the Entra ID directory.
+// The enterprise application:
+//   - Provides Teleport with OIDC authentication to Azure
+//   - Is given the permissions to access certain Microsoft Graph API endpoints for this tenant.
+//   - Provides SSO to the Teleport cluster via SAML.
+func SetupEnterpriseApp(ctx context.Context, proxyPublicAddr string, authConnectorName string) (string, string, error) {
+	var appID, tenantID string
+
+	tenantID, err := getTenantID()
+	if err != nil {
+		return appID, tenantID, trace.Wrap(err)
+	}
+
+	graphClient, err := createGraphClient()
+	if err != nil {
+		return appID, tenantID, trace.Wrap(err)
+	}
+
+	displayName := "Teleport" + " " + strings.TrimPrefix(proxyPublicAddr, "https://")
```

Comment:
nit: You can use `url.Url` to extract the hostname and make sure that the output is correct.